daos: add daos-serialize and daos-deserialize tools (#457)
hpc/mpifileutils: DAOS-Support.md

# DAOS Support

[DAOS](https://github.com/daos-stack/daos) is supported as a backend storage system in
dcp, dsync, and dcmp. Custom serialization and deserialization for DAOS
containers to and from a POSIX filesystem is provided with daos-serialize and

> daos-deserialize. These tools serialize a DAOS container and its associated
> metadata to an HDF5 file for storage on a POSIX filesystem. The container data

Reviewer: Rewording a bit to be specific about naming which tool to use for generating the HDF file.
```suggestion
daos-deserialize. A DAOS container and its associated
metadata can be serialized to an HDF5 file for storage on a POSIX filesystem using daos-serialize. The container data
```

Number of items that exist in both directories and have the same content: 1 (Src: 1 Dest: 1)
Number of items that exist in both directories and have different contents: 0 (Src: 0 Dest: 0)
```

## DAOS Serialization and Deserialization Examples

daos-serialize and daos-deserialize can be used on any type of DAOS container.
They are DAOS only tools that require HDF5.

daos-serialize will serialize a DAOS container to an HDF5 file. Depending on
the amount of data, multiple files may be written for each rank specified in the job.

daos-deserialize will deserialize or restore the HDF5 files written into a new
DAOS container, and a pool UUID to deserialize the data to must be specified.

#### Example One

Show the serialization of a DAOS container

```shell
$ mpirun -np 3 daos-serialize -v /$pool1/$cont1
Serializing Container to 0-7bf8037d-823f-4fa5-ac2a-c2cae8f81f57
Serializing Container to 1-7bf8037d-823f-4fa5-ac2a-c2cae8f81f57

> Serializing Container to 2-7bf8037d-823f-4fa5-ac2a-c2cae8f81f57

Author: @daltonbohning hmm I think maybe I could use <rank>_ I prefer to use it as a prefix though, because it is easier to specify each one if you are adding the files one by one to the command line with tab completion.

> Serializing Container to 2-7bf8037d-823f-4fa5-ac2a-c2cae8f81f57

Author: @daltonbohning I can also add "rank"

> Serializing Container to 2-7bf8037d-823f-4fa5-ac2a-c2cae8f81f57

Reviewer 2: @dsikich I'm not set on any way in particular, so we can always revisit this later if needed

> Serializing Container to 2-7bf8037d-823f-4fa5-ac2a-c2cae8f81f57

Reviewer: I haven't used HDF a lot, but is it convention to use an .hdf extension on those?  That might help people remember the file format, especially if other formats might be supported in the future.  I'm not stuck on that, though, and we can add it later if we think it'd be helpful.

Reviewer 2: @dsikich I think we probably should use a `.h5`  or `.hdf5` extension, per Mohamad's comment. Also, we should update the examples in this file with the `_rank` suffix you added.

Author: @daltonbohning this has been updated.

```

#### Example Two

Show the serialization of a DAOS container by specifying output directory

```shell
$ mpirun -np 3 daos-serialize -v -o serialize /$pool1/$cont1
Serializing Container to serialize/0-7bf8037d-823f-4fa5-ac2a-c2cae8f81f57
Serializing Container to serialize/1-7bf8037d-823f-4fa5-ac2a-c2cae8f81f57
Serializing Container to serialize/2-7bf8037d-823f-4fa5-ac2a-c2cae8f81f57
```
#### Example Three 

Show the deserialization of an HDF5 file

```shell
$ mpirun -np 3 daos-deserialize -v 0-7bf8037d-823f-4fa5-ac2a-c2cae8f81f57

> 1-7bf8037d-823f-4fa5-ac2a-c2cae8f81f57 2-7bf8037d-823f-4fa5-ac2a-c2cae8f81f57 

Author: @daltonbohning hmm I see, yeah that's a good idea we can do that
